src/main/java/org/broadinstitute/hellbender/tools/spark/sv/GatherSplitReadsSpark.java

```diff
@@ -0,0 +1,664 @@
+    // library fragment length.
+    // Discordant pairs aren't concordant, except that we're also ignoring reads with mates mapped to other contigs.
+    // Class assumes that all reads presented to it are mapped.
+    private static final class DiscordantPairDetector
+    {
+        private static final int FUNKY_PAIR_LOCUS_WIDTH = 100; // uncertainty in event locus for discordant pair
+        //TODO: Above should depend on the fragment size distribution
+
+        public EventLocus getLocusIfDiscordant( GATKRead read )
+        {
+            if ( read.mateIsUnmapped() )
+            {
+                int locus = read.isReverseStrand() ? read.getStart() - FUNKY_PAIR_LOCUS_WIDTH : read.getEnd();
+                return new EventLocus(locus,FUNKY_PAIR_LOCUS_WIDTH);
+            }
+            else if ( read.getContig() == read.getMateContig() )
+            {
+                if ( read.isReverseStrand() == read.mateIsReverseStrand() || !read.isProperlyPaired() )
+                {
+                    int locus = read.getFragmentLength() < 0 ? read.getStart() - FUNKY_PAIR_LOCUS_WIDTH : read.getEnd();
+                    return new EventLocus(locus,FUNKY_PAIR_LOCUS_WIDTH);
+                }
+            }
+
+            return null;
+        }
+    }
+
+    // A class that acts as a filter for a stream of reads.  It passes only those that are a part of a putative
+    // breakpoint cluster.
+    private static final class BreakpointClusterer implements Iterator<GATKRead>, Iterable<GATKRead>
+    {
+        private SplitReadDetector splitReadDetector;
+        private DiscordantPairDetector discordantPairDetector;
+        private final Iterator<GATKRead> inputIterator;
+        private String currentContig = null;
+        private long eventIdCounter = 0;
+        private final SortedMap<EventLocus,GATKRead> locMap = new TreeMap<>();
+        private Iterator<GATKRead> outputIterator = Collections.emptyIterator();
+
+        private static final int MIN_MATCH_LEN = 45; // minimum length of matched portion of an interesting alignment
+        private static final int MIN_MAPQ = 20; // minimum mapping quality of an interesting alignment
+        private static final int MAX_LOCUS_DIST = 300; // stale evidence distance, should be about the fragment length
+        private static final int MIN_EVIDENCE = 5; // minimum evidence count in a cluster
+
+        public BreakpointClusterer( final SplitReadDetector someSplitReadDetector,
+                                    final DiscordantPairDetector someDiscordantPairDetector,
+                                    final Iterator<GATKRead> someInputIterator )
+        {
+            splitReadDetector = someSplitReadDetector;
+            discordantPairDetector = someDiscordantPairDetector;
+            inputIterator = someInputIterator;
+        }
+
+        @Override
+        public boolean hasNext()
```

Comment:
typically hasNext() should not change the state of the iterator.  It seems like here if I call hasNext() repeatedly I'll run through the input iterator.